src/net_processing.cpp

```diff
@@ -2431,7 +2424,10 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             LOCK(cs_main);
             // Also always process if we requested the block explicitly, as we may
             // need it even though it is not a candidate for a new best tip.
-            forceProcessing |= MarkBlockAsReceived(hash);
+            // TODO: Only process if requested from this peer?
+            forceProcessing |= mmapBlocksInFlight.count(hash);
+            // Block is no longer in flight from this peer
+            MarkBlockAsNotInFlight(hash, pfrom->GetId());
```

Comment:
Probably makes sense for MarkBlockAsNotInFlight() to return whether it removed something for the peer? I think that would take care of the TODO as well.